compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java

```diff
@@ -372,23 +404,63 @@ public void visit(BLangSimpleVariableDef varDefNode) {
      * @return assignment statement created
      */
     private BLangAssignment createAssignment(BLangSimpleVariableDef varDefNode) {
-        // Create a map for the function or block node.
-        BLangBlockStmt blockStmt = (BLangBlockStmt) env.node;
-        if (blockStmt.mapSymbol == null) {
-            // create a map symbol
-            blockStmt.mapSymbol = createMapSymbol("$map$block$" + blockClosureMapCount, env);
-        }
+        BVarSymbol mapSymbol = createMapSymbolIfAbsent(env.node, blockClosureMapCount);
 
         // Add the variable to the created map.
-        BLangIndexBasedAccess accessExpr = ASTBuilderUtil.createIndexBasesAccessExpr(varDefNode.pos, varDefNode.type,
-                blockStmt.mapSymbol, ASTBuilderUtil.createLiteral(varDefNode.pos, symTable.stringType,
-                        varDefNode.var.name.value));
-        accessExpr.type = ((BMapType) blockStmt.mapSymbol.type).constraint;
+        BLangIndexBasedAccess accessExpr =
+                ASTBuilderUtil.createIndexBasesAccessExpr(varDefNode.pos, varDefNode.type, mapSymbol,
+                                                          ASTBuilderUtil
+                                                                  .createLiteral(varDefNode.pos, symTable.stringType,
+                                                                                 varDefNode.var.name.value));
+        accessExpr.type = ((BMapType) mapSymbol.type).constraint;
         accessExpr.lhsVar = true;
         // Written to: 'map["x"] = 8'.
         return ASTBuilderUtil.createAssignmentStmt(varDefNode.pos, accessExpr, varDefNode.var.expr);
     }
 
+    private BVarSymbol createMapSymbolIfAbsent(BLangNode node, int closureMapCount) {
+        if (node.getKind() == NodeKind.BLOCK_FUNCTION_BODY) {
+            return createMapSymbolIfAbsent((BLangBlockFunctionBody) node, closureMapCount);
+        } else if (node.getKind() == NodeKind.BLOCK) {
+            return createMapSymbolIfAbsent((BLangBlockStmt) node, closureMapCount);
+        } else if (node.getKind() == NodeKind.FUNCTION) {
+            return createMapSymbolIfAbsent((BLangFunction) node, closureMapCount);
+        }
+        return null;
```

Comment:
The name suggests that we create a map symbol always but the function may return null as well. Should this null be an assert ?

Reply:
`null` return is for the cases where a closure var map symbol is not applicable.

Reply:
To further clarify this, for nodes for which closures are not applicable, calling this method doesn't have any effect and we don't do anything with the return value. We use the return value returned by this method only in contexts where the closure map is applicable.